// devui/lib/Select/Select.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactSelect = require('react-select');

var _reactSelect2 = _interopRequireDefault(_reactSelect);

var _createStyledComponent = require('../utils/createStyledComponent');

var _createStyledComponent2 = _interopRequireDefault(_createStyledComponent);

var _styles = require('./styles');

var _styles2 = _interopRequireDefault(_styles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SelectContainer = (0, _createStyledComponent2.default)(_styles2.default, _reactSelect2.default);

var Select = function (_ref) {
  _inherits(Select, _ref);

  function Select() {
    _classCallCheck(this, Select);

    return _possibleConstructorReturn(this, (Select.__proto__ || Object.getPrototypeOf(Select)).apply(this, arguments));
  }

  _createClass(Select, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(SelectContainer, this.props);
    }
  }]);

  return Select;
}(_react.PureComponent || _react.Component);

exports.default = Select;


Select.propTypes = {
  autosize: _react.PropTypes.bool, // whether to enable autosizing or not
  clearable: _react.PropTypes.bool, // should it be possible to reset value
  disabled: _react.PropTypes.bool, // whether the Select is disabled or not
  isLoading: _react.PropTypes.bool, // whether the Select is loading externally or not
  menuMaxHeight: _react.PropTypes.number, // maximum css height for the opened menu of options
  multi: _react.PropTypes.bool, // multi-value input
  searchable: _react.PropTypes.bool, // whether to enable searching feature or not
  simpleValue: _react2.default.PropTypes.bool, // pass the value with label to onChange
  value: _react.PropTypes.any, // initial field value
  valueKey: _react.PropTypes.string, // path of the label value in option objects
  openOuterUp: _react.PropTypes.bool // value to control the opening direction
};

Select.defaultProps = { autosize: true, clearable: false, simpleValue: true, menuMaxHeight: 200 };
