// devui/lib/colorSchemes/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _default = require('./default');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_default).default;
  }
});

var _github = require('./github');

Object.defineProperty(exports, 'github', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_github).default;
  }
});

var _atomOneDark = require('./atom-one-dark');

Object.defineProperty(exports, 'atomOneDark', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_atomOneDark).default;
  }
});

var _dracula = require('./dracula');

Object.defineProperty(exports, 'dracula', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dracula).default;
  }
});

var _irBlack = require('./ir-black');

Object.defineProperty(exports, 'iRBlack', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_irBlack).default;
  }
});

var _macintosh = require('./macintosh');

Object.defineProperty(exports, 'macintosh', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_macintosh).default;
  }
});

var _materia = require('./materia');

Object.defineProperty(exports, 'materia', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_materia).default;
  }
});

var _oceanicNext = require('./oceanic-next');

Object.defineProperty(exports, 'oceanicNext', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_oceanicNext).default;
  }
});

var _phd = require('./phd');

Object.defineProperty(exports, 'phD', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_phd).default;
  }
});

var _pico = require('./pico');

Object.defineProperty(exports, 'pico', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_pico).default;
  }
});

var _solarFlare = require('./solar-flare');

Object.defineProperty(exports, 'solarFlare', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_solarFlare).default;
  }
});

var _spacemacs = require('./spacemacs');

Object.defineProperty(exports, 'spacemacs', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_spacemacs).default;
  }
});

var _unikitty = require('./unikitty');

Object.defineProperty(exports, 'unikitty', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_unikitty).default;
  }
});

var _woodland = require('./woodland');

Object.defineProperty(exports, 'woodland', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_woodland).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
