// violet-paginator/lib/decorators/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withRecordProps = exports.violetPaginator = exports.tabulateLean = exports.tabulate = exports.sort = exports.stretch = exports.paginate = exports.flip = exports.decorate = undefined;

var _decorate2 = require('./decorate');

var _decorate3 = _interopRequireDefault(_decorate2);

var _flip2 = require('./flip');

var _flip3 = _interopRequireDefault(_flip2);

var _paginate2 = require('./paginate');

var _paginate3 = _interopRequireDefault(_paginate2);

var _stretch2 = require('./stretch');

var _stretch3 = _interopRequireDefault(_stretch2);

var _sort2 = require('./sort');

var _sort3 = _interopRequireDefault(_sort2);

var _tabulate2 = require('./tabulate');

var _tabulate3 = _interopRequireDefault(_tabulate2);

var _tabulateLean2 = require('./tabulateLean');

var _tabulateLean3 = _interopRequireDefault(_tabulateLean2);

var _violetPaginator2 = require('./violetPaginator');

var _violetPaginator3 = _interopRequireDefault(_violetPaginator2);

var _withRecordProps2 = require('./withRecordProps');

var _withRecordProps3 = _interopRequireDefault(_withRecordProps2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.decorate = _decorate3.default;
exports.flip = _flip3.default;
exports.paginate = _paginate3.default;
exports.stretch = _stretch3.default;
exports.sort = _sort3.default;
exports.tabulate = _tabulate3.default;
exports.tabulateLean = _tabulateLean3.default;
exports.violetPaginator = _violetPaginator3.default;
exports.withRecordProps = _withRecordProps3.default;
